proxy/src/control/discovery.rs

```diff
@@ -188,7 +189,12 @@ where
                 let service = self.bind.bind(&addr).map_err(|_| ())?;
 
                 Ok(Async::Ready(Change::Insert(addr, service)))
-            }
+            },
+            Update::ChangeMetadata(addr) => {
+                let service = self.bind.bind(&addr).map_err(|_| ())?;
+
+                Ok(Async::Ready(Change::Insert(addr, service)))
```

Comment:
Let's document the reasoning here, because this is very unclear. In fact it seems wrong to me, but I might be misunderstanding it.

Is this really supposed to be `Change::Insert` when this is an update and not an insertion? If so, why?

My understanding is that this shouldn't result in any change being sent at all since we don't forward the metadata upstream, and instead this line should be replaced by `/// TODO: Send metadata update notification`.

Otherwise, isn't this case and the `Update::Insert()` case exactly the same? It seems like they should be in the same match arm if they're doing the same thing.

Reply:
@briansmith 
> Is this really supposed to be `Change::Insert` when this is an update and not an insertion? If so, why?

This is `Change::Insert` because `tower-discover` lacks a concept of a metadata update. We do have to re-bind the service, as `self.bind.bind()` is responsible for adding the `Labeled` middleware to the bound service. A metadata change (currently) implies a change in labels, so we need to change the `Labeled` middleware wrapping the bound service, and to my knowledge there's no way to do that besides binding a new service for the address.

 > Otherwise, isn't this case and the `Update::Insert()` case exactly the same? It seems like they should be in the same match arm if they're doing the same thing.

I thought these _were_ collapsed into one match arm in this branch, but it looks like I actually only made that change in #661: https://github.com/runconduit/conduit/pull/661/files#diff-3d7563188396c0392342b192778deb15L186

Reply:
> This is Change::Insert because tower-discover lacks a concept of a metadata update. We do have to re-bind the service, as self.bind.bind() is responsible for adding the Labeled middleware to the bound service. A metadata change (currently) implies a change in labels, so we need to change the Labeled middleware wrapping the bound service, and to my knowledge there's no way to do that besides binding a new service for the address.

OK, but does everything tolerate a "duplicate insert"? That is, is Tower's semantics for `Change::Insert` really an "upsert" i.e. "Insert or Update"? Is there any test that things work correctly on a duplicate insert?

It would be good to confirm with @carllerche that this is the best way of handling this.

Reply:
@briansmith thanks for pointing this out, I did some research and it looks like this is the wrong behaviour. `tower-balance` is the consumer of these discovery changes, and it looks like it ignores duplicate `Change::Insert`s: https://github.com/tower-rs/tower/blob/11b591b6e037ae671bc8da755cfb3095a3f83568/tower-balance/src/lib.rs#L149-L156

I think the correct thing to do in this case is to send a `Change::Remove` _followed by_ a `Change::Insert` for the re-bound service.

Reply:
Update: I asked @olix0r briefly about this, and he said that the right thing to do is probably to change `tower-discover` to have "upsert" behaviour on `Change::Insert` events. I'll open an issue and PR against `tower` to make that change.

Reply:
I added a comment in 108b847dafea87fb58a6bf9a7343bc8cdabb3c47 describing the changes I will make in  #661.